include/util/static_rtree.hpp

```diff
@@ -166,15 +167,245 @@ class StaticRTree
+        default:
+            PackWithHilbert(input_data_vector, tree_node_filename, leaf_node_filename);
+        }
+    }
+
+    // Construct a packed STR-RTree with the Luetengger-Edgington-Lopez approach
+    // STR performs better than Hilbert for road-network-like data where the
+    // distribution is only mildly biased
+    void PackWithSTR(const std::vector<EdgeDataT> &input_data_vector,
+                     const std::string &tree_node_filename,
+                     const std::string &leaf_node_filename)
+    {
+        util::Log() << "Packing with STR";
+        auto copy = const_cast<std::vector<EdgeDataT> &>(input_data_vector);
+        //        copy.reserve(input_data_vector.size());
+        // std::copy(input_data_vector.begin(), input_data_vector.end(), copy);
+
+        // Sort by longitude coordinate of centroid for each segment
+        // TODO: this means calculating centroids a lot, this could be moved into
+        // a separate single pass
+        util::Log() << "Sorting leaves by centroid longitude";
+        tbb::parallel_sort(
+            copy.begin(), copy.end(), [this](const EdgeDataT &a, const EdgeDataT &b) {
+                auto a_centroid = coordinate_calculation::centroid(m_coordinate_list[a.u],
+                                                                   m_coordinate_list[a.v]);
+                auto b_centroid = coordinate_calculation::centroid(m_coordinate_list[b.u],
+                                                                   m_coordinate_list[b.v]);
+                return a_centroid.lon < b_centroid.lon;
+
+            });
+
+        // Now, break into even-sized vertical slices, and sort each slice by latitude
+        // TODO: make sure this isn't 0
+        const std::size_t leaf_K = std::sqrt(copy.size() / LEAF_NODE_SIZE);
+        util::Log() << "leaf_K is " << leaf_K;
+        tbb::parallel_for(tbb::blocked_range<uint64_t>(0, copy.size(), leaf_K),
+                          [&copy, this](const tbb::blocked_range<uint64_t> &range) {
+                              util::Log() << "Sorting vertical leaf slice " << range.begin() << "-"
+                                          << range.end();
+                              tbb::parallel_sort(
+                                  copy.begin() + range.begin(),
+                                  copy.begin() + range.end(),
+                                  [this](const EdgeDataT &a, const EdgeDataT &b) {
+                                      auto a_centroid = coordinate_calculation::centroid(
+                                          m_coordinate_list[a.u], m_coordinate_list[a.v]);
+                                      auto b_centroid = coordinate_calculation::centroid(
+                                          m_coordinate_list[b.u], m_coordinate_list[b.v]);
+                                      return a_centroid.lat < b_centroid.lat;
+                                  });
+                          });
+
+        // open leaf file
```

Comment:
Code below is a duplication. Better to avoid this by dropping `HilbertInputElement` (it will require computation of Hilbert values in comparator, but reduce memory consumption :sunglasses:) and use a generalized `PackingPolicy` for soring.